xmtp_mls/src/groups/mod.rs

```diff
@@ -211,13 +224,93 @@ where
         ))
     }
 
+    fn process_own_message(
+        &self,
+        conn: &mut DbConnection,
+        intent: StoredGroupIntent,
+        openmls_group: &mut OpenMlsGroup,
+        provider: &XmtpOpenMlsProvider,
+        message: ProtocolMessage,
+        envelope_timestamp_ns: u64,
+    ) -> Result<(), MessageProcessingError> {
+        if intent.state == IntentState::Committed {
+            return Ok(());
+        }
+        debug!(
+            "[{}]processing own message for intent {} / {:?}",
+            self.client.account_address(),
+            intent.id,
+            intent.kind
+        );
+        let message_epoch = message.epoch();
+        let group_epoch = openmls_group.epoch();
+
+        // TODO: Figure out if we really need to do this for application messages
+        // With the max_past_epochs flag this likely isn't necessary
+        if group_epoch != message_epoch {
+            log::warn!(
+                "wrong epoch. expected {}. received {}",
+                group_epoch,
+                message_epoch
+            );
+            // All other intent kinds create pending commits. They need to be cleared
+            if intent.kind != IntentKind::SendMessage {
+                openmls_group.clear_pending_commit();
+            }
+            // Revert intent to ToPublish state
+            self.client
+                .store
+                .set_group_intent_to_publish(conn, intent.id)?;
+
+            return Err(MessageProcessingError::WrongEpoch {
+                message_epoch,
+                group_epoch,
+            });
+        }
+
+        match intent.kind {
+            IntentKind::AddMembers | IntentKind::RemoveMembers | IntentKind::KeyUpdate => {
+                debug!("[{}] merging pending commit", self.client.account_address());
+                openmls_group.merge_pending_commit(provider)?;
```

Comment:
Some off-the-cuff thoughts here, not necessarily for this PR immediately:

1. [merge_staged_commit](https://openmls.tech/openmls/doc/openmls/group/struct.MlsGroup.html#method.merge_staged_commit) clears any pending commits - we probably actually need to handle requeuing any pending intents/commits on the receive flow when merging other people's commits. 
2. Because of that, I'm not sure if the `group_epoch != message_epoch` above is necessary or not
3. No rush, but I think we need to handle at least some of the errors on this line in the same way you're handling the `group_epoch != message_epoch` case - in fact maybe that check is done inside this method

Reply:
1. You're totally right. We should also be explicitly checking that there is a pending commit to be merged here and erroring if there isn't. Will fix.
2. I'm down to move this logic into an error branch of `merge_pending_commit`. Does feel cleaner to just let openmls tell us when there is a problem instead of trying to predict it.

Reply:
OK. Changed the logic up to not check for epoch matches and instead handle errors or `merge_pending_commit` and also explicitly return an error if there is no pending commit.

I looked at the code and tested it out and was a little surprised to find that `merge_pending_commit` just returns `OK(())` when there is no pending commit. That seems like an odd choice, but it's how OpenMLS was designed.

```rust
    pub fn merge_pending_commit<KeyStore: OpenMlsKeyStore>(
        &mut self,
        provider: &impl OpenMlsProvider<KeyStoreProvider = KeyStore>,
    ) -> Result<(), MergePendingCommitError<KeyStore::Error>> {
        match &self.group_state {
            MlsGroupState::PendingCommit(_) => {
                let old_state = mem::replace(&mut self.group_state, MlsGroupState::Operational);
                if let MlsGroupState::PendingCommit(pending_commit_state) = old_state {
                    self.merge_staged_commit(provider, (*pending_commit_state).into())?;
                }
                Ok(())
            }
            MlsGroupState::Inactive => Err(MlsGroupStateError::UseAfterEviction)?,
            MlsGroupState::Operational => Ok(()),
        }
    }
```